app/models/concerns/has_recent_items.rb

```diff
@@ -13,16 +13,26 @@ def recent_item_types
       self.class.recent_item_types
     end
 
-    def recent_items
-      viewable_recent_item_links.order(updated_at: :desc).preload(:item).map(&:item).compact
+    def recent_items(limit: 10)
+      items = recent_item_types.values.flat_map do |klass|
+        viewable_recent_items(klass, limit: limit)
+      end
+      items.sort_by { |item, date| [date, item.id] }.map(&:first).reverse!.take(limit)
     end
 
-    def viewable_recent_item_links
-      ids = recent_item_types.values.map do |item_class|
-        recent_item_links.where(item_type: item_class.name, item_id: item_class.viewable_by(self).pluck(:id)).pluck(:id)
-      end.flatten
+    # @returns [item, last_updated]
+    protected def viewable_recent_items(item_klass, limit:)
+      links_by_item = recent_item_links.
+        where(item_type: item_klass.sti_name).
+        order(updated_at: :desc).
+        limit(limit).
+        to_a.index_by(&:item_id)
 
-      recent_item_links.where(id: ids)
+      # restrict by visibility
+      item_klass.viewable_by(self).where(id: links_by_item.keys).map do |item|
+        link = links_by_item.fetch(item.id)
+        [item, link.updated_at]
+      end
```

Comment:
If you can't see any of the most recent 10, aren't you going to get nothing back from this?  (The limit is imposed on finding the most recent items before the view ability is enforced.)

Reply:
It's a good point. My thought was to avoid loading the ids for every item the user has ever viewed just to keep the top 10. I think it is probably okay to keep the limit where it is with the following assumptions:
a) it's pretty rare for someone to lose access to a recent client or project
b) the impact of not seeing those recent clients/projects is minor since this just a convenience and navigation shortcut

If that's not true I can move the limit to viewable items. It probably won't make much of a difference in most situations.

Reply:
I think it's fine.. I agree its a convenience, so we should focus on it being fast over being complete in an edge case